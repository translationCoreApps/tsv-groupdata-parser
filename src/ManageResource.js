import fs from 'fs-extra'
import path from 'path-extra'

class ManageResource {
  constructor(sourcePath) {
    this.resource = {}
    this.sourcePath = sourcePath
  }

  loadChapter(bookId, chapter) {
    const bookDir = path.join(this.sourcePath, bookId);
    if (fs.existsSync(bookDir)) {
      const chapterFile = path.join(bookDir, chapter + '.json');
      if (fs.existsSync(chapterFile) && !this.resource[chapter]) {
        const chapterObject = fs.readFileSync(chapterFile)
        this.resource[chapter] = chapterObject;
      } else {
        console.error(`${chapterFile}, path does not exist`)
      }
    } else {
      console.error(`${bookDir}, path does not exist`)
    }
  }

  getResource() {
    return this.resource;
  }

  getVerse(chapter, verse) {
    return this.resource[chapter][verse];
  }
}

export default ManageResource;
