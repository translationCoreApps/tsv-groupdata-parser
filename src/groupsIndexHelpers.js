import fs from 'fs-extra'
import path from 'path-extra'

export const generateGroupsIndex = (tnCategoriesPath, taCategoriesPath) => {
  const categorizedGroupsIndex = {
    discourse: [],
    numbers: [],
    figures: [],
    culture: [],
    grammar: [],
    other: []
  }

  const isDirectory = (item) => fs.lstatSync(path.join(tnCategoriesPath, item)).isDirectory()
  const categories = fs.readdirSync(tnCategoriesPath).filter(isDirectory)

  categories.forEach(categoryName => {
    const booksPath = path.join(tnCategoriesPath, categoryName, 'groups')
    const books = fs.readdirSync(booksPath).filter(item => item !== '.DS_Store');
    books.forEach(bookid => {
      const groupDataPath = path.join(booksPath, bookid)
      const groupDataFiles = fs.readdirSync(groupDataPath).filter((filename) => path.extname(filename) === '.json');
      groupDataFiles.forEach(groupDataFile => {
        const filePath = path.join(groupDataPath, groupDataFile)
        const groupId = groupDataFile.replace('.json', '')
        const groupData = fs.readJsonSync(filePath)

        if (groupData.length > 0) {
          const taArticleCategory = getArticleCategory(groupData[0].contextId.occurrenceNote)
          const fileName = groupId + '.md'
          const articlePath = path.join(taCategoriesPath, taArticleCategory, fileName)
          const groupName = getGroupName(articlePath)
          const groupIndexItem = getGroupIndex(groupId, groupName);

          // Only add the groupIndexItem if it isnt already in the category's groups index.
          if (!categorizedGroupsIndex[categoryName].some(e => e.id === groupIndexItem.id)) {
            categorizedGroupsIndex[categoryName].push(groupIndexItem)// adding group Index Item
          }
        }
      })
    })
  })

  fs.outputJsonSync('categorizedGroupsIndex.json', categorizedGroupsIndex, {spaces:2})

  return categorizedGroupsIndex;
}

const getArticleCategory = (occurrenceNote) => {
  const cutEnd = occurrenceNote.search("rc://en/ta/man/");
  const taArticleCategory = (occurrenceNote.substr(0, 0) + occurrenceNote.substr(cutEnd + 1))
    .replace("c://en/ta/man/", "")
    .split("/")[0]

  return taArticleCategory;
}

const getGroupName = (articlePath) => {
  const articleFile = fs.readFileSync(articlePath, 'utf8');
  // get the article's first line and remove #'s and spaces from beginning/end
  return articleFile.split('\n')[0].replace(/(^\s*#\s*|\s*#\s*$)/gi, '');
}

const getGroupIndex = (groupId, groupName) => {
  return {
    id: groupId,
    name: groupName
  }
}
