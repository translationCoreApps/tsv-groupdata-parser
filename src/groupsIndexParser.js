import fs from 'fs-extra';
import path from 'path-extra';
import { getGroupName } from './helpers/resourcesHelpers';

export const generateGroupsIndex = (tnCategoriesPath, taCategoriesPath) => {
  const categorizedGroupsIndex = {
    discourse: [],
    numbers: [],
    figures: [],
    culture: [],
    grammar: [],
    other: [],
  };
  let error = false;

  const isDirectory = item => fs.lstatSync(path.join(tnCategoriesPath, item)).isDirectory();
  const categories = fs.readdirSync(tnCategoriesPath).filter(isDirectory);

  categories.forEach(categoryName => {
    const booksPath = path.join(tnCategoriesPath, categoryName, 'groups');
    const books = fs.readdirSync(booksPath).filter(item => item !== '.DS_Store');

    books.forEach(bookid => {
      const groupDataPath = path.join(booksPath, bookid);
      const groupDataFiles = fs.readdirSync(groupDataPath).filter(filename => path.extname(filename) === '.json');
      let taArticleCategory;
      let groupName;

      groupDataFiles.forEach(groupDataFile => {
        try {
          const filePath = path.join(groupDataPath, groupDataFile);
          const groupId = groupDataFile.replace('.json', '');
          const groupData = fs.readJsonSync(filePath);
          taArticleCategory = null;
          groupName = null;

          if (groupData.length > 0) {
            taArticleCategory = getArticleCategory(groupData[0].contextId.occurrenceNote, groupId);

            if (taArticleCategory) {
              const fileName = groupId + '.md';
              const articlePath = path.join(taCategoriesPath, taArticleCategory, fileName);
              groupName = getGroupName(articlePath);
              const groupIndexItem = getGroupIndex(groupId, groupName);

              // Only add the groupIndexItem if it isn't already in the category's groups index.
              if (!categorizedGroupsIndex[categoryName].some(e => e.id === groupIndexItem.id)) {
                categorizedGroupsIndex[categoryName].push(groupIndexItem); // adding group Index Item
              }
            }
          }
        } catch (e) {
          console.error(`generateGroupsIndex() - error processing entry: bookid: ${bookid}, groupDataFile: ${groupDataFile}, taArticleCategory: ${taArticleCategory}, groupName: ${groupName}`);
          error = true;
        }
      });
    });
  });

  if (error) {
    throw new Error(`generateGroupsIndex() - error processing index`);
  }

  return categorizedGroupsIndex;
};

/**
 * Gets the category of the given groupId from the links in the occurrenceNote
 * @param {string} occurrenceNote
 * @param {string} groupId
 * @returns {string} the matched category
 */
export const getArticleCategory = (occurrenceNote, groupId) => {
  if (occurrenceNote && groupId) {
    const pattern = '(?<=rc:\\/\\/[^\\/]+\\/ta\\/man\\/)[^\\/]+?(?=\\/' + groupId + ')';
    const categoryRE = new RegExp(pattern);
    const match = occurrenceNote.match(categoryRE);

    if (match) {
      return match[0];
    }
  }
  return null;
};

const getGroupIndex = (groupId, groupName) => ({
  id: groupId,
  name: groupName,
});
