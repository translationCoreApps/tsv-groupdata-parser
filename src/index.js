export {
  tsvToGroupData
} from './tsvToGroupData'
export {
  getCategoryForGroupId,
  categorizeGroupData,
  tNotesCategories
} from './tNoteGroupIdCategorization'
export {
  formatAndSaveGroupData,
  saveGroupsIndex
} from './filesystemHelpers'
export {
  generateGroupsIndex
} from './groupsIndexHelpers'
