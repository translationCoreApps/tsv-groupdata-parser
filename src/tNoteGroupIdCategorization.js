/**
 * Hard coded list of tNotes categories
 */
export const tNotesCategories = {
  lexical: {
    "translate-bweight": "Measures and Weights",
    "translate-numbers": "Numbers",
    "translate-fraction": "Fractions",
    "translate-ordinal": "Ordinal Numbers"
  },
  figures: {
    "figs-idiom": "Idiom",
    "figs-irony": "Irony",
    "figs-metaphor": "Metaphor",
    "figs-rquestions": "Rhetorical Questions",
    "figs-simile": "Simile",
    "figs-apostrophe": "apostrophe",
    "figs-euphemism": "Euphemism",
    "figs-hendiadys": "Hendiadys",
    "figs-hyperbole": "Hyperbole",
    "figs-litotes": "litotes",
    "figs-merism": "Merism",
    "figs-metonymy": "Metonymy",
    "figs-parallelism": "Parallelism",
    "figs-personification": "Personification",
    "figs-synecdoche": "Synecdoche",
    "figs-ellipsis": "Ellipsis",
    "figs-doublenegatives": "Double negatives"
  },
  cultural: {
    "translate-symaction": "Symbolic Actions",
    "writing-symlanguage": "Symbolic Language",
    "translate-unknown": "Unknowns",
    "figs-explicit": "Implicit and Explicit Information"
  },
  morphological: {
    "figs-you": "Forms of You",
    "figs-we": "Forms of We",
    "figs-they": "Forms of They",
    "figs-activepassive": "Active or Passive",
    "figs-gendernotations": "Gender",
    "figs-pronouns": "Pronouns",
    "writing-connectingwords": "Connecting Words",
  }
}

/**
 * Gets the category name of a groupId. Returns other as a categpory name
 * if the groupId is not found in the tNotes categories in the list above.
 * @param {string} groupId tA article slug.
 * @returns {string} The category name of a groupId: lexical,
 * figures, cultural, morphological or other.
 */
export const getCategoryForGroupId = (groupId) => {
  let result = 'other';

  Object.keys(tNotesCategories)
    .forEach((category) => {
      if (tNotesCategories[category][groupId]) result = category
    })

  return result;
}

/**
 * Categorizes the group data into five categories: lexical,
 * figures, cultural, morphological and other.
 * @param {Object} groupData Object of groupids that contain an
 * array of contextIds for each groupId.
 * @returns The group data categorized into five categories:
 * lexical, figures, cultural, morphological and other.
 * 
 * {
    lexical: {},
    figures: {
      'figs-metaphor': []// array of groupId objects
    },
    cultural: {},
    morphological: {
      'figs-activepassive': []// array of groupId objects
    },
    other: {
      'figs-abstractnouns': []// array of groupId objects
    }
  }
 */
export const categorizeGroupData = groupData => {
  const categorizedGroupData = {
    lexical: {},
    figures: {
      'figs-metaphor': []
    },
    cultural: {
      'figs-explicit': []
    },
    morphological: {
      'figs-activepassive': []
    },
    other: {
      'figs-abstractnouns': [],
    }
  }

  Object.keys(groupData).forEach(groupId => {
    const category = getCategoryForGroupId(groupId)
    categorizedGroupData[category][groupId] = groupData[groupId]
  })

  return categorizedGroupData;
}