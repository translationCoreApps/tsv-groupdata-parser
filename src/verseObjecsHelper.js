
export function verseObjectsToString(verseObjects) {
  return verseObjects.map(verseObject => verseObject.text).join('');
};
