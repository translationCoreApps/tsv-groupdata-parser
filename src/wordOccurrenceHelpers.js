import stringTokenizer from 'string-punctuation-tokenizer'
// constants
const ELLIPSES = '...'

function countStringInArray(array, string) {
  return array.filter(item => item == string).length;
}

function substrOccurrencesInQuote(quote, substr, substrIndex) {
  const quoteSubstrings = stringTokenizer.tokenizeWithPunctuation(quote)
  const leftSubstrs = quoteSubstrings.slice(0, substrIndex)
  return countStringInArray(leftSubstrs, substr)
}

function getQuoteOmittedString(quote, verseString) {
  const quoteChunks = quote.split(ELLIPSES)
  let missingWordsIndices = [];

  quoteChunks.forEach((quoteChunk, index) => {
    let quoteChunkSubStrIndex

    // Get quote chunk closer to next quote chunk
    if (index < (quoteChunks.length - 1)) {
      const nextQuoteChunk = quoteChunks[index + 1]
      const nextChunkIndex = verseString.indexOf(nextQuoteChunk)
      if (nextChunkIndex) {
        const strBeforeNextQuote = verseString.substring(0, nextChunkIndex)
        quoteChunkSubStrIndex = strBeforeNextQuote.lastIndexOf(quoteChunk)
      }
    } else {
      quoteChunkSubStrIndex = verseString.indexOf(quoteChunk);
    }

    missingWordsIndices.push(quoteChunkSubStrIndex + (index === 0 ? quoteChunk.length : 0))
  })

  const [beginIndex, endIndex] = missingWordsIndices;
  const missingStringChunk = verseString.slice(beginIndex, endIndex)

  return missingStringChunk;
}

/**
 * search for previous occurrences of word to get occurrence
 * for this instance
 * @param {string} verseString - The string to search in.
 * @param {string} substr - The sub string to search for.
 * @param {string} quote - the orig language quote.
 * @param {number} substrIndex - substring index number.
 * @param {string} quoteOmittedString - string omitted by ellipses
 */
function getWordOccurrence(verseString, substr, quote, substrIndex, quoteOmittedString) {
  const quoteWithOmittedString = quote.includes(ELLIPSES) ?
    quote.replace(ELLIPSES, quoteOmittedString) : quote
  const quoteSubStrIndex = verseString.indexOf(quoteWithOmittedString);
  const previousStr = verseString.substring(0, quoteSubStrIndex);
  const previousStrs = stringTokenizer.tokenizeWithPunctuation(previousStr)
  let previousOccurrences = 0;

  for (let i = 0; i <= quoteSubStrIndex; i++) {
    const stringItem = previousStrs[i];
    if (stringItem === substr) {
      previousOccurrences++;
    }
  }

  let occurrence = ++previousOccurrences

  // if substr is found in quote more than once
  if ((quoteWithOmittedString.split(new RegExp(substr, 'gi')).length - 1) > 1) {
    if (quote.includes(ELLIPSES)) {
      quote = quote.replace(/\.../g, '')
    }
    let previousSubstrOccurrences = substrOccurrencesInQuote(quote, substr, substrIndex)
    occurrence = ++previousSubstrOccurrences
  }

  return occurrence
}


export function getWordOccurrencesForQuote(quote, verseString) {
  let words = [],
      quoteOmittedString = '',
      cleanedQuote = quote

  if (quote.includes(ELLIPSES)) {
    cleanedQuote = quote.replace(/\.../g, '')
    quoteOmittedString = getQuoteOmittedString(quote, verseString)
  }

  const substrings = stringTokenizer.tokenizeWithPunctuation(cleanedQuote)

  substrings.forEach((substring, index) => {
    const word = {
      word: substring,
      occurrence: getWordOccurrence(verseString, substring, quote, index, quoteOmittedString),
    }

    words.push(word)
  });

  return words
}
