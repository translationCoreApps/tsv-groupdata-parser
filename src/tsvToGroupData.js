import tsvtojson from 'tsvtojson'
import { categorizeGroupData } from './tNoteGroupIdCategorization'
import ManageResource from './helpers/ManageResourceAPI'
import { getWordOccurrencesForQuote } from './helpers/wordOccurrenceHelpers'

/**
 * Parses a book tN TSVs and returns an object holding the lists of group ids.
 * @param {string} filepath path to tsv file.
 * @param {string} toolName tC tool name.
 * @param {object} params When it includes { categorized: true }
 * then it returns the object organized by tn article category.
 * @param {string} originalBiblePath path to original bible
 * @returns an object with the lists of group ids which each
 * includes an array of groupsdata.
 */
export const tsvToGroupData = async (filepath, toolName, params = {}, originalBiblePath) => {
  const groupData = {}
  const tsvObjects = await tsvtojson(filepath)
  const { Book: bookId } = tsvObjects[0] || {}
  const resourceApi = new ManageResource(originalBiblePath, bookId.toLowerCase())

  tsvObjects.map(tsvItem => {
    if (tsvItem.SupportReference && tsvItem.OrigQuote) {
      tsvItem.SupportReference = cleanGroupId(tsvItem.SupportReference)
      tsvItem.OccurrenceNote = cleanArticleLink(tsvItem.OccurrenceNote)
      const chapter = parseInt(tsvItem.Chapter, 10)
      const verse = parseInt(tsvItem.Verse, 10)
      const verseString = resourceApi.getVerseString(chapter, verse)

      if (groupData[tsvItem.SupportReference]) {
        groupData[tsvItem.SupportReference].push(generateGroupDataItem(tsvItem, toolName, verseString))
      } else {
        groupData[tsvItem.SupportReference] = [generateGroupDataItem(tsvItem, toolName, verseString)]
      }
    }
  })

  return params && params.categorized ? categorizeGroupData(groupData) : groupData
}

/**
 * Cleans an incorrectly formatted group id.
 * @param {string} groupId group id string that was posibly incorrectly formatted.
 * @returns correctly formatted group id.
 */
export const cleanGroupId = groupId => {
  const subStrings = groupId.replace(/translate:|translate\//gi, '').split(/[_\/:]/g)

  if (subStrings.length === 1) {
    return subStrings[0]
  } else if (subStrings.length === 2) {
    return subStrings.join('-')
  } else {
    return groupId
  }
}

/**
 * Finds incorrectly formatted tA links and fixes them.
 * @param {string} occurrenceNote occurrence Note.
 * @returns {string} occurrenceNote with clean/fixed tA links.
 */
export const cleanArticleLink = occurrenceNote => {
  let noteWithFixedLink = ''
  const linkSubstring = 'rc://en/ta/man/'
  const cutEnd = occurrenceNote.search(linkSubstring)
  const groupId = (occurrenceNote.substr(0, 0) + occurrenceNote.substr(cutEnd + 1)).replace('c://en/ta/man/', '').replace(']])', '')
  const stringFirstPart = occurrenceNote.slice(0, cutEnd)

  if (groupId.includes(linkSubstring)) {
    // handle multiple links in the same note.
    const joint = ' and '
    const multipleLinksSubstrings = groupId.split(joint)
    const lastItem = multipleLinksSubstrings.length - 1
    let goodLinks = ''

    multipleLinksSubstrings.forEach((substring, index) => {
      const linkString = substring.replace('[[rc://en/ta/man/', '').replace(']]', '')
      const cleanedGropId = cleanGroupId(linkString)

      if (index === 0) {
        goodLinks = goodLinks + `rc://en/ta/man/translate/${cleanedGropId}]]` + joint
      } else if (index !== lastItem) {
        goodLinks = goodLinks + `[[rc://en/ta/man/translate/${cleanedGropId}]]` + joint
      } else if (index === lastItem) {
        goodLinks = goodLinks + `[[rc://en/ta/man/translate/${cleanedGropId}]])`
      }
    })
    noteWithFixedLink = stringFirstPart + goodLinks
  } else {
    // only one link in the note
    const cleanedGropId = cleanGroupId(groupId)
    const goodLink = `rc://en/ta/man/translate/${cleanedGropId}]])`
    noteWithFixedLink = stringFirstPart + goodLink
  }

  return noteWithFixedLink
}

/**
 * Returns the formatted groupData item for a given tsv item.
 * @param {object} tsvItem tsv item.
 * @param {string} toolName tool name.
 * @returns {object} groupData item.
 */
const generateGroupDataItem = (tsvItem, toolName, verseString) => {
  const { OrigQuote = '' } = tsvItem
  // if quote has more than one word get word occurrences
  const quote = OrigQuote.trim().split(' ').length > 1 ? getWordOccurrencesForQuote(OrigQuote, verseString) : OrigQuote

  return {
    comments: false,
    reminders: false,
    selections: false,
    verseEdits: false,
    contextId: {
      occurrenceNote: tsvItem.OccurrenceNote || '',
      reference: {
        bookId: tsvItem.Book.toLowerCase() || '',
        chapter: parseInt(tsvItem.Chapter, 10) || '',
        verse: parseInt(tsvItem.Verse, 10) || '',
      },
      tool: toolName || '',
      groupId: tsvItem.SupportReference || '',
      quote,
      glQuote: tsvItem.GLQuote || '',
      occurrence: parseInt(tsvItem.Occurrence, 10) || 1,
    },
  }
}
