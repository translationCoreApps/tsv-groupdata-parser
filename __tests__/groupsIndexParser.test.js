jest.unmock('fs-extra')
import { generateGroupsIndex, getArticleCategory } from '../src/groupsIndexParser'

describe('tests groupsIndexParser.generateGroupsIndex()', () => {
  test('returns an object with all the tn categories, each one with their groupsIndex', async () => {
    const tnCategoriesPath = '__tests__/fixtures/resources/en/translationHelps/translationNotes/v14'
    const taCategoriesPath = '__tests__/fixtures/resources/en/translationHelps/translationAcademy/v10'
    const result = generateGroupsIndex(tnCategoriesPath, taCategoriesPath)
    expect(result).toMatchSnapshot()
  })
})

describe('tests groupsIndexParser.getArticleCategory()', () => {
  test('returns the category found in an occurrenceNote', async () => {
    const groupId = 'figs-metaphor'
    const category = 'translate'
    const occurrenceNote = 'This is my note (See: [Metaphor](rc://en/ta/man/' + category + '/' + groupId + '))'
    expect(getArticleCategory(occurrenceNote, groupId)).toBe(category)
  })

  test('returns the category found in an occurrenceNote as the second link', async () => {
    const groupId = 'figs-metaphor'
    const category = 'translate'
    const occurrenceNote = 'This is my note (See: [Spelling](rc://en/ta/man/checking/spelling) and [Metaphor](rc://en/ta/man/' + category + '/' + groupId + '))'
    expect(getArticleCategory(occurrenceNote, groupId)).toBe(category)
  })

  test('returns the category found in a Hindi occurrenceNote', async () => {
    const groupId = 'figs-metaphor'
    const category = 'translate'
    const occurrenceNote = 'पौलूस परमेश्वर के सन्देश की इस तरह बात करता है जैसे कि वह कोई वस्तु हो जिसे लोगों को दृश्य रूप में दिखाया जा सकता हो। वैकल्पिक अनुवाद: “उसने मुझे अपना सन्देश समझने के लिए प्रेरित किया” (देखें: [[rc://hi/ta/man/' + category + '/' + groupId + ']])'
    expect(getArticleCategory(occurrenceNote, groupId)).toBe(category)
  })

  test('returns no categories for a note with no links', async () => {
    const groupId = 'some-group-id'
    const occurrenceNote = 'This is my note'
    expect(getArticleCategory(occurrenceNote, groupId)).toBe(null)
  })

  test('returns no categories for a empty occurrenceNote', async () => {
    const groupId = null
    const occurrenceNote = null
    expect(getArticleCategory(occurrenceNote, groupId)).toBe(null)
  })
})
