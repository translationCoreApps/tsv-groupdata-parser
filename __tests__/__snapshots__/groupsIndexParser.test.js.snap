// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests groupsIndexParser.generateGroupsIndex() returns an object with all the tn categories, each one with their groupsIndex 1`] = `
Object {
  "culture": Array [
    Object {
      "id": "figs-explicit",
      "name": "Assumed Knowledge and Implicit Information",
    },
    Object {
      "id": "translate-bdistance",
      "name": "Biblical Distance",
    },
    Object {
      "id": "translate-bmoney",
      "name": "Biblical Money",
    },
    Object {
      "id": "translate-hebrewmonths",
      "name": "Hebrew Months",
    },
    Object {
      "id": "translate-names",
      "name": "How to Translate Names",
    },
    Object {
      "id": "translate-symaction",
      "name": "Symbolic Action",
    },
    Object {
      "id": "translate-unknown",
      "name": "Translate Unknowns",
    },
    Object {
      "id": "writing-symlanguage",
      "name": "Symbolic Language",
    },
    Object {
      "id": "translate-bvolume",
      "name": "Biblical Volume",
    },
    Object {
      "id": "translate-bweight",
      "name": "Biblical Weight",
    },
    Object {
      "id": "figs-go",
      "name": "Go and Come",
    },
  ],
  "discourse": Array [
    Object {
      "id": "figs-exclamations",
      "name": "Exclamations",
    },
    Object {
      "id": "figs-quotesinquotes",
      "name": "Quotes Within Quotes",
    },
    Object {
      "id": "translate-versebridge",
      "name": "Verse Bridges",
    },
    Object {
      "id": "writing-background",
      "name": "Background Information",
    },
    Object {
      "id": "writing-newevent",
      "name": "Introduction of a New Event",
    },
    Object {
      "id": "writing-poetry",
      "name": "Poetry",
    },
    Object {
      "id": "figs-quotations",
      "name": "Direct and Indirect Quotations",
    },
    Object {
      "id": "writing-proverbs",
      "name": "Proverbs",
    },
    Object {
      "id": "figs-events",
      "name": "Order of Events",
    },
    Object {
      "id": "writing-connectingwords",
      "name": "Connecting Words",
    },
    Object {
      "id": "figs-parables",
      "name": "Parables",
    },
    Object {
      "id": "writing-endofstory",
      "name": "End of Story",
    },
    Object {
      "id": "writing-participants",
      "name": "Introduction of New and Old Participants",
    },
    Object {
      "id": "figs-pastforfuture",
      "name": "Predictive Past",
    },
    Object {
      "id": "writing-quotations",
      "name": "Quotations and Quote Margins",
    },
    Object {
      "id": "figs-imperative",
      "name": "Imperatives - Other Uses",
    },
    Object {
      "id": "figs-sentences",
      "name": "Sentence Structure",
    },
    Object {
      "id": "figs-declarative",
      "name": "Statements - Other Uses",
    },
    Object {
      "id": "writing-intro",
      "name": "Types of Writing",
    },
    Object {
      "id": "figs-exmetaphor",
      "name": "Extended Metaphor",
    },
  ],
  "figures": Array [
    Object {
      "id": "figs-doublet",
      "name": "Doublet",
    },
    Object {
      "id": "figs-ellipsis",
      "name": "Ellipsis",
    },
    Object {
      "id": "figs-euphemism",
      "name": "Euphemism",
    },
    Object {
      "id": "figs-hyperbole",
      "name": "Hyperbole and Generalization",
    },
    Object {
      "id": "figs-idiom",
      "name": "Idiom",
    },
    Object {
      "id": "figs-merism",
      "name": "Merism",
    },
    Object {
      "id": "figs-metaphor",
      "name": "Metaphor",
    },
    Object {
      "id": "figs-metonymy",
      "name": "Metonymy",
    },
    Object {
      "id": "figs-parallelism",
      "name": "Parallelism",
    },
    Object {
      "id": "figs-personification",
      "name": "Personification",
    },
    Object {
      "id": "figs-rquestion",
      "name": "Rhetorical Question",
    },
    Object {
      "id": "figs-simile",
      "name": "Simile",
    },
    Object {
      "id": "figs-synecdoche",
      "name": "Synecdoche",
    },
    Object {
      "id": "figs-apostrophe",
      "name": "Apostrophe",
    },
    Object {
      "id": "figs-doublenegatives",
      "name": "Double Negatives",
    },
    Object {
      "id": "figs-hendiadys",
      "name": "Hendiadys",
    },
    Object {
      "id": "figs-irony",
      "name": "Irony",
    },
    Object {
      "id": "figs-litotes",
      "name": "Litotes",
    },
  ],
  "grammar": Array [
    Object {
      "id": "figs-123person",
      "name": "First, Second, or Third Person",
    },
    Object {
      "id": "figs-abstractnouns",
      "name": "Abstract Nouns",
    },
    Object {
      "id": "figs-activepassive",
      "name": "Active or Passive",
    },
    Object {
      "id": "figs-exclusive",
      "name": "Exclusive and Inclusive \\"We\\"",
    },
    Object {
      "id": "figs-nominaladj",
      "name": "Nominal Adjectives",
    },
    Object {
      "id": "figs-rpronouns",
      "name": "Reflexive Pronouns",
    },
    Object {
      "id": "figs-you",
      "name": "Forms of You",
    },
    Object {
      "id": "figs-hypo",
      "name": "Hypothetical Situations",
    },
    Object {
      "id": "figs-inclusive",
      "name": "Inclusive \\"We\\"",
    },
    Object {
      "id": "figs-possession",
      "name": "Possession",
    },
    Object {
      "id": "figs-distinguish",
      "name": "Distinguishing versus Informing or Reminding",
    },
    Object {
      "id": "figs-genericnoun",
      "name": "Generic Noun Phrases",
    },
    Object {
      "id": "figs-pronouns",
      "name": "Pronouns",
    },
    Object {
      "id": "figs-gendernotations",
      "name": "When Masculine Words Include Women",
    },
  ],
  "numbers": Array [
    Object {
      "id": "translate-numbers",
      "name": "Numbers",
    },
    Object {
      "id": "translate-ordinal",
      "name": "Ordinal Numbers",
    },
    Object {
      "id": "translate-fraction",
      "name": "Fractions",
    },
  ],
  "other": Array [
    Object {
      "id": "translate-textvariants",
      "name": "Textual Variants",
    },
    Object {
      "id": "guidelines-sonofgodprinciples",
      "name": "Translating Son and Father",
    },
    Object {
      "id": "translate-manuscripts",
      "name": "Original Manuscripts",
    },
    Object {
      "id": "translate-transliterate",
      "name": "Copy or Borrow Words",
    },
  ],
}
`;

exports[`tests groupsIndexParser.generateGroupsIndex() test error handling 1`] = `
[Error: generateGroupsIndex() - error processing index: error processing entry: bookid: gal, groupDataFile: translate-numbers.json, taArticleCategory: translate, groupName: null: Error: ENOENT: no such file or directory, open 'translate/translate-numbers.md'
error processing entry: bookid: tit, groupDataFile: figs-hypo.json, taArticleCategory: translate, groupName: null: Error: ENOENT: no such file or directory, open 'translate/figs-hypo.md']
`;
